backends/aerospike.go

```diff
@@ -13,59 +14,105 @@ import (
+type AerospikeBackend struct {
 	cfg     config.Aerospike
-	client  *as.Client
+	client  AerospikeDB
 	metrics *metrics.Metrics
 }
 
-func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *Aerospike {
+func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *AerospikeBackend {
 	client, err := as.NewClient(cfg.Host, cfg.Port)
 	if err != nil {
-		log.Fatalf("Error creating Aerospike backend: %v", err)
-		panic("Aerospike failure. This shouldn't happen.")
+		log.Fatalf("%v", formatAerospikeError(err).Error())
+		panic("AerospikeBackend failure. This shouldn't happen.")
 	}
 	log.Infof("Connected to Aerospike at %s:%d", cfg.Host, cfg.Port)
 
-	return &Aerospike{
+	return &AerospikeBackend{
 		cfg:     cfg,
-		client:  client,
+		client:  &AerospikeDBClient{client},
 		metrics: metrics,
 	}
 }
 
-func (a *Aerospike) Get(ctx context.Context, key string) (string, error) {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Get(ctx context.Context, key string) (string, error) {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, key)
 	if err != nil {
-		return "", err
+		return "", formatAerospikeError(err, "GET")
 	}
-	rec, err := a.client.Get(nil, asKey, "value")
+	rec, err := a.client.Get(asKey)
 	if err != nil {
-		return "", err
+		return "", formatAerospikeError(err, "GET")
 	}
 	if rec == nil {
-		return "", errors.New("client.Get returned a nil record. Is aerospike configured properly?")
+		return "", formatAerospikeError(errors.New("Nil record"), "GET")
 	}
 	a.metrics.RecordExtraTTLSeconds(float64(rec.Expiration))
-	return rec.Bins[binValue].(string), nil
+
+	value, found := rec.Bins[binValue]
+	if !found {
+		return "", formatAerospikeError(errors.New("No 'value' bucket found"), "GET")
+	}
+
+	str, isString := value.(string)
+	if !isString {
+		return "", formatAerospikeError(errors.New("Unexpected non-string value found"), "GET")
+	}
+
+	return str, nil
 }
 
-func (a *Aerospike) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, key)
 	if err != nil {
-		return err
+		return formatAerospikeError(err, "PUT")
 	}
+
 	if ttlSeconds == 0 {
 		ttlSeconds = a.cfg.DefaultTTL
 	}
-	bins := as.BinMap{
-		binValue: value,
+
+	bins := as.BinMap{binValue: value}
+	policy := &as.WritePolicy{Expiration: uint32(ttlSeconds)}
+
+	if err := a.client.Put(policy, asKey, bins); err != nil {
+		return formatAerospikeError(err, "PUT")
 	}
-	err = a.client.Put(&as.WritePolicy{
-		Expiration: uint32(ttlSeconds),
-	}, asKey, bins)
+
+	return nil
+}
+
+func formatAerospikeError(err error, caller ...string) error {
 	if err != nil {
-		return err
+		msg := "Aerospike"
+		if len(caller) == 1 && len(caller[0]) > 0 {
```

Comment:
If we're only ever expecting a single caller in the arguments then why have it as a variable string. argument?

Reply:
I thought it'd be "nicer" if we could pass it one or two arguments when needed. I thought it'd look nices to sometimes do:
```
formatAerospikeError(err,"METHOD")
```
And others simply:
```
formatAerospikeError(err)
```

Reply:
So we can accomplish the above by passing in just a single string argument as well and you can send an empty string in cases where you don't want a `caller` to be added to the error message. I am okay with having variable string arguments as well but in that case I'd suggest making the code a little generic by changing this if condition to a for statement that loops over the `caller` string slice and appends them to the error message

Reply:
Corrected